app/osparc.py

```diff
@@ -1,71 +1,159 @@
+            temp_input_file.seek(0)
 
-            if status.state != "PUBLISHED":
-                raise SimulationException("the simulation job could not be submitted")
+            try:
+                files_api = osparc.FilesApi(api_client)
 
-            # Wait for the simulation job to be complete (or to fail).
+                input_file = files_api.upload_file(temp_input_file.name)
+            except ApiException as e:
+                raise SimulationException(
+                    f"the solver input file could not be uploaded ({e})")
 
-            while True:
-                status = solvers_api.inspect_job(solver.id, solver.version, job.id)
+            temp_input_file.close()
 
-                if status.progress == 100:
-                    break
+        # Create the simulation job with the job inputs that matches our
+        # simulation type.
 
-                sleep(1)
+        solvers_api = osparc.SolversApi(api_client)
 
-            status = solvers_api.inspect_job(solver.id, solver.version, job.id)
+        try:
+            solver = solvers_api.get_solver_release(
+                solver_name, data["solver"]["version"])
+        except ApiException as e:
+            raise SimulationException(
+                f"the requested solver could not be retrieved ({e})")
+
+        if solver_name == OPENCOR_SOLVER:
+            job_inputs = {
+                "model_url": data["opencor"]["model_url"],
+                "config_file": config_file
+            }
+        else:
+            if has_solver_input:
+                data["osparc"]["job_inputs"][data["solver"]["input"]["name"]] = input_file
+
+            job_inputs = data["osparc"]["job_inputs"]
+
+        job = solvers_api.create_job(
+            solver.id,
+            solver.version,
+            osparc.JobInputs(job_inputs)
+        )
+
+        # Start the simulation job.
+
+        status = solvers_api.start_job(solver.id, solver.version, job.id)
+
+        if status.state != "PUBLISHED":
+            raise SimulationException(
+                "the simulation job could not be submitted")
+
+        res = {
+            "status": "ok",
+            "data": {
+                "job_id": job.id,
+                "solver": {
+                    "name": solver.id,
+                    "version": solver.version
+                }
+            }
+        }
+    except SimulationException as e:
+        res = {
+            "status": "nok",
+            "description": e.args[0] if len(e.args) > 0 else "unknown"
+        }
+
+    return res
+
+
+def check_simulation(data):
+    try:
+        # Check whether the simulation has completed (or failed).
+
+        api_client = osparc.ApiClient(osparc.Configuration(
+            host=Config.OSPARC_API_URL,
+            username=Config.OSPARC_API_KEY,
+            password=Config.OSPARC_API_SECRET
+        ))
+        solvers_api = osparc.SolversApi(api_client)
+        job_id = data["job_id"]
```

Comment:
Do we need to check the keys from data? Or are they always correct?

Reply:
`data` comes from oSPARC and, unless they change their JSON schema (which would break many things for them, I would imagine), we should be fine. Also, this is the reason we do all of that stuff in a `try...except` statement. If something goes wrong, we report an error.

Reply:
The data is checked in line 1145 in main.py.

Reply:
> The data is checked on line 1145 in main.py.

Oh yes, I had forgotten that I had done that check. :)

Reply:
Thanks @alan-wu! I missed that